src/charm.py

```diff
@@ -53,43 +53,32 @@ def __init__(self, *args: Any) -> None:
             strip_prefix=True,
         )
         self.framework.observe(self.on.config_changed, self._on_config_changed)
-        self.framework.observe(self.on.reset_instance_action, self._reset_instance_action)
+        self.framework.observe(self.on.reset_instance_action, self._on_reset_instance_action)
 
     def _on_config_changed(self, event: ops.HookEvent) -> None:
         """Handle changed configuration.
 
         Args:
             event: Event triggering after config is changed.
         """
-        try:
-            self._charm_state = CharmState.from_charm(charm=self)
-        except CharmConfigInvalidError as exc:
-            self.model.unit.status = ops.BlockedStatus(exc.msg)
-            return
         container = self.unit.get_container(SYNAPSE_CONTAINER_NAME)
         if not container.can_connect():
             event.defer()
             self.unit.status = ops.WaitingStatus("Waiting for pebble")
             return
-        server_name_configured = self._synapse.server_name_configured(container)
-        if (
-            server_name_configured is not None
-            and server_name_configured != self._charm_state.server_name
-        ):
-            msg = (
-                f"server_name {self._charm_state.server_name} is different from the existing one"
-                f" {server_name_configured}."
-                " Please revert the config or"
-                " run the action reset-instance if you to erase the existing instance and start a"
-                " new one."
-            )
-            self.model.unit.status = ops.BlockedStatus(msg)
-            return
-        self.model.unit.status = ops.MaintenanceStatus("Configuring Synapse")
         try:
+            self._charm_state = CharmState.from_charm(charm=self)
```

Comment:
In most cases we do this in the `__init__` function, would that work here?

Reply:
I'm glad that you asked because I wanted to ask you and @gregory-schiano about that and I forgot to do it yesterday.
While comparing the server_name I noticed that the state isn't updated when the config changes. From what is described [here](https://juju.is/docs/sdk/create-a-minimal-kubernetes-charm) I was expecting that the init would be called on config-changed event so I don't know what I'm missing here.

Reply:
Well this is already done in the `__init__`see L41
So I think this is not needed

Reply:
If I dont call it again, the state is not updated with the new config :|

Reply:
This is very very strange.
Normally each hook is executed in a new instance BUT I read that this is not guaranteed (but never good clear answers on when this could happen)
Is it systematic ?

Reply:
Hmmm the deployed charm works as expected (juju config -> init -> config-changed) but the unit tests using "harness.update_config({"server_name": server_name_changed})" dont call init again.

Reply:
https://github.com/canonical/operator/issues/736

Reply:
Ah indeed, that's something we realised with Weii too while working on Flask charm

Reply:
I added a new fixture called harness_server_name_changed_fixture that will force the charm to be initialized before the config-chaged event :)